File: daffodil-japi/src/main/scala/edu/illinois/ncsa/daffodil/japi/infoset/Infoset.scala
/* Copyright (c) 2017 Tresys Technology, LLC. All rights reserved.
 *
 * Developed by: Tresys Technology, LLC
 *               http://www.tresys.com
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal with
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:
 *
 *  1. Redistributions of source code must retain the above copyright notice,
 *     this list of conditions and the following disclaimers.
 *
 *  2. Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimers in the
 *     documentation and/or other materials provided with the distribution.
 *
 *  3. Neither the names of Tresys Technology, nor the names of its contributors
 *     may be used to endorse or promote products derived from this Software
 *     without specific prior written permission.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH THE
 * SOFTWARE.
 */

package edu.illinois.ncsa.daffodil.japi.infoset

import edu.illinois.ncsa.daffodil.infoset.{ InfosetOutputter => SInfosetOutputter }
import edu.illinois.ncsa.daffodil.infoset.{ InfosetInputter => SInfosetInputter }
import edu.illinois.ncsa.daffodil.infoset.{ ScalaXMLInfosetOutputter => SScalaXMLInfosetOutputter }
import edu.illinois.ncsa.daffodil.infoset.{ XMLTextInfosetOutputter => SXMLTextInfosetOutputter }
import edu.illinois.ncsa.daffodil.infoset.{ JsonInfosetOutputter => SJsonInfosetOutputter }
import edu.illinois.ncsa.daffodil.infoset.{ NullInfosetOutputter => SNullInfosetOutputter }
import edu.illinois.ncsa.daffodil.infoset.{ JDOMInfosetOutputter => SJDOMInfosetOutputter }
import edu.illinois.ncsa.daffodil.infoset.{ ScalaXMLInfosetInputter => SScalaXMLInfosetInputter }
import edu.illinois.ncsa.daffodil.infoset.{ XMLTextInfosetInputter => SXMLTextInfosetInputter }
import edu.illinois.ncsa.daffodil.infoset.{ JsonInfosetInputter => SJsonInfosetInputter }
import edu.illinois.ncsa.daffodil.infoset.{ JDOMInfosetInputter => SJDOMInfosetInputter }
import edu.illinois.ncsa.daffodil.util.MaybeBoolean
import edu.illinois.ncsa.daffodil.infoset.InfosetInputterEventType
// TODO: Not sure about the access to internal infoset implementation details.
// Should API users have this deep access to our internal infoset?
import edu.illinois.ncsa.daffodil.infoset.DISimple
import edu.illinois.ncsa.daffodil.infoset.DIComplex
import edu.illinois.ncsa.daffodil.infoset.DIArray

/**
 * Abstract class used to determine how the infoset representation should be
 * input from a call to DataProcessor#unparse. This uses a Cursor API, such
 * that each call to advance/inspect must update a cursor value, minimizing
 * allocations. Callers of advance/inspect are expected to copy out any
 * information from advanceAccessor and inspectAccessor if they need to retain
 * the information after a call to advance/inspect.
 **/
abstract class InfosetInputter extends SInfosetInputter {
  /**
   * Return the current infoset inputter event type
   */
  def getEventType(): InfosetInputterEventType 

  /**
   * Get the local name of the current event. This will only be called when the
   * current event type is StartElement.
   */
  def getLocalName(): String

  /**
   * Get the namespace of the current event. This will only be called when the
   * current event type is StartElement. If the InfosetInputter does not
   * support namespaces, this shoud return null. This may return null to
   * represent no namespaces.
   */
  def getNamespaceURI(): String

  /**
   * Get the content of a simple type. This will only be called when the
   * current event type is StartElement and the element is a simple type. If
   * the event contains complex data, it is an error and should throw
   * NonTextFoundInSimpleContentException. If the element does not have any
   * simple content, this should return either null or the empty string.
   */
  def getSimpleText(): String

  /**
   * Determine if the current event is nilled. This will only be called when
   * the current event type is StartElement. Return MaybeBoolean.Nope if no
   * nil property is set, which implies the element is not nilled. Return
   * MaybeBoolean(false) if the nil property is set, but it is set to false.
   * Return MaybeBoolean(true) if the nil property is set to true.
   */
  def isNilled(): MaybeBoolean

  /**
   * Return true if there are remaining events. False otherwise.
   */
  def hasNext(): Boolean

  /**
   * Move the internal state to the next event.
   */
  def next(): Unit
}

/**
 * Abstract class used to determine how the infoset representation should be
 * output from a call to DataProcessor#parse. The Daffodil core will call
 * the various methods of this class in an order appropriate to create an
 * infoset representation.
 *
 * Classes that extend InfosetOutputter are not guaranteed to be thread-safe.
 **/
abstract class InfosetOutputter extends SInfosetOutputter {

  /**
   * Reset the internal state of this InfosetOutputter. This should be called
   * inbetween calls to the parse method.
   */
  def reset(): Unit // call to reuse these. When first constructed no reset call is necessary.

  /**
   * Called by Daffodil internals to signify the beginning of the infoset.
   *
   * @return true on sucess, false if there was an error and Daffodil should stop all
   *         future calls to the InfosetOutputter
   */
  def startDocument(): Boolean

  /**
   * Called by Daffodil internals to signify the end of the infoset.
   *
   * @return true on sucess, false if there was an error and Daffodil should stop all
   *         future calls to the InfosetOutputter
   */
  def endDocument(): Boolean

  /**
   * Called by Daffodil internals to signify the beginning of a simple element.
   *
   * @param diSimple the simple element that is started. Various fields of
   *                 DISimple can be accessed to determine things like the
   *                 value, nil, name, namespace, etc.
   * @return true on sucess, false if there was an error and Daffodil should stop all
   *         future calls to the InfosetOutputter
   */

  def startSimple(diSimple: DISimple): Boolean

  /**
   * Called by Daffodil internals to signify the end of a simple element.
   *
   * @param diSimple the simple element that is ended. Various fields of
   *                 DISimple can be accessed to determine things like the
   *                 value, nil, name, namespace, etc.
   * @return true on sucess, false if there was an error and Daffodil should stop all
   *         future calls to the InfosetOutputter
   */
  def endSimple(diSimple: DISimple): Boolean

  /**
   * Called by Daffodil internals to signify the beginning of a complex element.
   *
   * @param diComplex the complex element that is started. Various fields of
   *                  DIComplex can be accessed to determine things like the
   *                  nil, name, namespace, etc.
   * @return true on sucess, false if there was an error and Daffodil should stop all
   *         future calls to the InfosetOutputter
   */
  def startComplex(diComplex: DIComplex): Boolean

  /**
   * Called by Daffodil internals to signify the end of a complex element.
   *
   * @param diComplex the complex element that is ended. Various fields of
   *                  DIComplex can be accessed to determine things like the
   *                  nil, name, namespace, etc.
   * @return true on sucess, false if there was an error and Daffodil should stop all
   *         future calls to the InfosetOutputter
   */
  def endComplex(diComplex: DIComplex): Boolean

  /**
   * Called by Daffodil internals to signify the beginning of an array of elements.
   *
   * @param diArray the array that is started. Various fields of
   *                DIArray can be accessed to determine things like the
   *                name, namespace, etc.
   * @return true on sucess, false if there was an error and Daffodil should stop all
   *         future calls to the InfosetOutputter
   */
  def startArray(diArray: DIArray): Boolean

  /**
   * Called by Daffodil internals to signify the end of an array of elements.
   *
   * @param diArray the array that is ended. Various fields of
   *                DIArray can be accessed to determine things like the
   *                name, namespace, etc.
   * @return true on sucess, false if there was an error and Daffodil should stop all
   *         future calls to the InfosetOutputter
   */
  def endArray(diArray: DIArray): Boolean
}

/**
 * The below classes are all just proxies of the classes implemented in the
 * core daffodil code. The reason for this is purely for documentation. When
 * only generate javadoc on the daffodil-japi subproject. By having this proxy
 * classes, we can document these classes and have a small and clean javadoc.
 */

 
/**
 * Output the infoset as a scala.xml.Node
 *
 * @param showFormatInfo add additional properties to each scala.xml.Node for debug purposes
 */
class ScalaXMLInfosetOutputter(showFormatInfo: Boolean = false)
  extends InfosetOutputterProxy {

  override val infosetOutputter = new SScalaXMLInfosetOutputter(showFormatInfo)

  def getResult(): scala.xml.Node = infosetOutputter.getResult()
}

/**
 * Output the infoset as XML Text, written to a java.io.Writer
 *
 * @param writer the java.io.Writer to write the XML text to
 * @param pretty enable or disable pretty printing. Pretty printing will only
 *               inserts indentation and newlines where it will not affect the
 *               content of the XML.
 */
class XMLTextInfosetOutputter(writer: java.io.Writer, pretty: Boolean = true)
  extends InfosetOutputterProxy {
  
  override val infosetOutputter = new SXMLTextInfosetOutputter(writer, pretty)
}

/**
 * Output the infoset as json text, written to a java.io.Writer
 *
 * @param writer the java.io.Writer to write the json text to
 * @param pretty enable or disable pretty printing. Pretty printing will only
 *               inserts indentation and newlines where it will not affect the
 *               content of the json.
 */
class JsonInfosetOutputter(writer: java.io.Writer, pretty: Boolean = true)
  extends InfosetOutputterProxy {

  override val infosetOutputter = new SJsonInfosetOutputter(writer, pretty)  
}

/**
 * Output the infoset as a jdom Document
 */
class JDOMInfosetOutputter()
  extends InfosetOutputterProxy {

  override val infosetOutputter = new SJDOMInfosetOutputter()

  def getResult(): org.jdom2.Document = infosetOutputter.getResult()
}

/**
 * Ignore all infoset output
 */
class NullInfosetOutputter()
  extends InfosetOutputterProxy {

  override val infosetOutputter = new SNullInfosetOutputter()
}


/**
 * Read in an infoset in the form of a scala.xml.Node
 *
 * @param node the infoset in the form of a scala.xml.Node
 */
class ScalaXMLInfosetInputter(node: scala.xml.Node)
  extends InfosetInputterProxy {

  override val infosetInputter = new SScalaXMLInfosetInputter(node)
}

/**
 * Read in an infoset in the form of XML text from a java.io.Reader
 *
 * @param reader the java.io.Reader to read the XML text from
 */
class XMLTextInfosetInputter(reader: java.io.Reader)
  extends InfosetInputterProxy {
  
  override val infosetInputter = new SXMLTextInfosetInputter(reader)
}

/**
 * Read in an infoset in the form of json text from a java.io.Reader
 *
 * @param reader the java.io.Reader to read the json text from
 */
class JsonInfosetInputter(reader: java.io.Reader)
  extends InfosetInputterProxy {

  override val infosetInputter = new SJsonInfosetInputter(reader)
}

/**
 * Read in an infoset in the form of jdom2 Document
 *
 * @param document the infoset in the form of a jdom2 Document
 */
class JDOMInfosetInputter(document: org.jdom2.Document)
  extends InfosetInputterProxy {

  override val infosetInputter = new SJDOMInfosetInputter(document)
}

/* A proxy for existing infoset inputters */
abstract class InfosetInputterProxy extends InfosetInputter {

  protected val infosetInputter: SInfosetInputter

  override def getEventType() = infosetInputter.getEventType()
  override def getLocalName() = infosetInputter.getLocalName()
  override def getNamespaceURI() = infosetInputter.getNamespaceURI()
  override def getSimpleText() = infosetInputter.getSimpleText()
  override def hasNext() = infosetInputter.hasNext()
  override def isNilled() = infosetInputter.isNilled()
  override def next() = infosetInputter.next()
  override lazy val supportsNamespaces = infosetInputter.supportsNamespaces

  override def fini = infosetInputter.fini
}

/* A proxy for existing infoset outputters */
abstract class InfosetOutputterProxy extends InfosetOutputter {

  protected val infosetOutputter: SInfosetOutputter

  override def reset(): Unit = infosetOutputter.reset()
  override def startDocument(): Boolean = infosetOutputter.startDocument()
  override def endDocument(): Boolean = infosetOutputter.endDocument()
  override def startSimple(diSimple: DISimple): Boolean = infosetOutputter.startSimple(diSimple)
  override def endSimple(diSimple: DISimple): Boolean = infosetOutputter.endSimple(diSimple)
  override def startComplex(diComplex: DIComplex): Boolean = infosetOutputter.startComplex(diComplex)
  override def endComplex(diComplex: DIComplex): Boolean = infosetOutputter.endComplex(diComplex)
  override def startArray(diArray: DIArray): Boolean = infosetOutputter.startArray(diArray)
  override def endArray(diArray: DIArray): Boolean = infosetOutputter.endArray(diArray)
}

File: daffodil-core/src/main/java/org/apache/daffodil/api/infoset/Infoset.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.daffodil.api.infoset;

import org.apache.daffodil.api.InputSourceDataInputStream;
import org.apache.daffodil.runtime1.infoset.JsonInfosetOutputter;
import org.apache.daffodil.runtime1.infoset.XMLTextInfosetOutputter;
import scala.xml.Node;

import java.io.InputStream;
import java.io.OutputStream;
import java.nio.ByteBuffer;

/**
 * Contains factory methods for InfosetInputters and Outputters
 */
public class Infoset {
  /**
   * private constructor
   */
  private Infoset() {
  }

  /**
   * {@link InfosetOutputter} to build an infoset represented as a scala.xml.Node
   *
   * @return InfosetOutputter
   */
  public static ScalaXMLInfosetOutputter getScalaXMLInfosetOutputter() {
    return new org.apache.daffodil.runtime1.infoset.ScalaXMLInfosetOutputter(false);
  }

  /**
   * {@link InfosetOutputter} to build an infoset represented as XML written to a java.io.OutputStream
   * <p>
   * Output the infoset as XML Text, written to a java.io.OutputStream
   *
   * @param os     the java.io.OutputStream to write the XML text to
   * @param pretty enable or disable pretty printing. Pretty printing will only
   *               insert indentation and newlines where it will not affect the
   *               content of the XML.
   * @return InfosetOutputter
   */
  public static InfosetOutputter getXMLTextInfosetOutputter(OutputStream os, boolean pretty) {
    return new XMLTextInfosetOutputter(os, pretty, XMLTextEscapeStyle.Standard, false);
  }

  /**
   * {@link InfosetOutputter} to build an infoset represented as XML written to a java.io.OutputStream
   * <p>
   * Output the infoset as XML Text, written to a java.io.OutputStream
   *
   * @param os                 the java.io.OutputStream to write the XML text to
   * @param pretty             enable or disable pretty printing. Pretty printing will only
   *                           insert indentation and newlines where it will not affect the
   *                           content of the XML.
   * @param xmlTextEscapeStyle determine whether to wrap values of elements of type
   *                           xs:string in CDATA tags in order to preserve
   *                           whitespace.
   * @return InfosetOutputter
   */
  public static InfosetOutputter getXMLTextInfosetOutputter(OutputStream os, boolean pretty, XMLTextEscapeStyle xmlTextEscapeStyle) {
    return new XMLTextInfosetOutputter(os, pretty, xmlTextEscapeStyle, false);
  }

  /**
   * {@link InfosetOutputter} to build an infoset represented as JSON written to a java.io.OutputStream
   * Output the infoset as json text, written to a java.io.OutputStream
   *
   * @param os     the java.io.OutputStream to write the json text to
   * @param pretty enable or disable pretty printing. Pretty printing will only
   *               insert indentation and newlines where it will not affect the
   *               content of the json.
   * @return InfosetOutputter
   */
  public static InfosetOutputter getJsonInfosetOutputter(OutputStream os, boolean pretty) {
    return new JsonInfosetOutputter(os, pretty);
  }

  /**
   * {@link InfosetOutputter} to build an infoset represented as an org.jdom2.Document
   *
   * @return InfosetOutputter
   */
  public static JDOMInfosetOutputter getJDOMInfosetOutputter() {
    return new org.apache.daffodil.runtime1.infoset.JDOMInfosetOutputter();
  }

  /**
   * {@link InfosetOutputter} to build an infoset represented as an org.w3c.dom.Document
   *
   * @return InfosetOutputter
   */
  public static W3CDOMInfosetOutputter getW3CDOMInfosetOutputter() {
    return new org.apache.daffodil.runtime1.infoset.W3CDOMInfosetOutputter();
  }

  /**
   * {@link InfosetOutputter} that does not build an infoset representation, ignoring
   * all {@link InfosetOutputter} events
   *
   * @return InfosetOutputter
   * \
   */
  public static InfosetOutputter getNullInfosetOutputter() {
    return new org.apache.daffodil.runtime1.infoset.NullInfosetOutputter();
  }

  /**
   * {@link InfosetInputter} to read an infoset represented as a scala.xml.Node
   *
   * @param node the scala.xml.Node infoset
   * @return InfosetInputter
   */
  public static InfosetInputter getScalaXMLInfosetInputter(Node node) {
    return new org.apache.daffodil.runtime1.infoset.ScalaXMLInfosetInputter(node);
  }

  /**
   * {@link InfosetInputter} to read an infoset represented as XML from a java.io.InputStream
   * <p>
   * Read in an infoset in the form of XML text from a java.io.InputStream
   *
   * @param is the java.io.InputStream to read the XML text from
   * @return InfosetInputter
   */
  public static InfosetInputter getXMLTextInfosetInputter(InputStream is) {
    return new org.apache.daffodil.runtime1.infoset.XMLTextInfosetInputter(is);
  }

  /**
   * {@link InfosetInputter} to read an infoset represented as JSON from a java.io.InputStream
   * <p>
   * Read in an infoset in the form of json text from a java.io.InputStream
   *
   * @param is the java.io.InputStream to read the json text from
   * @return InfosetInputter
   */
  public static InfosetInputter getJsonInfosetInputter(InputStream is) {
    return new org.apache.daffodil.runtime1.infoset.JsonInfosetInputter(is);
  }

  /**
   * {@link InfosetInputter} to read an infoset represented as an org.jdom2.Document
   *
   * @param document the org.jdom2.Document infoset
   * @return InfosetInputter
   */
  public static InfosetInputter getJDOMInfosetInputter(org.jdom2.Document document) {
    return new org.apache.daffodil.runtime1.infoset.JDOMInfosetInputter(document);
  }

  /**
   * {@link InfosetInputter} to read an infoset represented as an org.w3c.dom.Document
   *
   * @param document the org.w3c.dom.Document infoset. Note that w3c
   *                 Documents are not guaranteed to be thread-safe, even if all
   *                 users only read/traverse it. It is up to the user to ensure
   *                 that the Document passed into the W3CDOMInfosetInputter is
   *                 not read or written by other threads while the
   *                 W3CDOMInfosetInputter has access to it.
   * @return InfosetInputter
   */
  public static InfosetInputter getW3CDOMInfosetInputter(org.w3c.dom.Document document) {
    return new org.apache.daffodil.runtime1.infoset.W3CDOMInfosetInputter(document);
  }

  /**
   * Create an InputSourceDataInputStream from a java.io.InputStream
   *
   * @param is input stream to create from
   * @return InputSourceDataInputStream from a java.io.InputStream
   */
  public static InputSourceDataInputStream getInputSourceDataInputStream(InputStream is) {
    return org.apache.daffodil.io.InputSourceDataInputStream.apply(is);
  }

  /**
   * Create an InputSourceDataInputStream from a java.nio.ByteBuffer
   *
   * @param bb byte buffer to create from
   * @return InputSourceDataInputStream from a java.nio.ByteBuffer
   */
  public static InputSourceDataInputStream getInputSourceDataInputStream(ByteBuffer bb) {
    return org.apache.daffodil.io.InputSourceDataInputStream.apply(bb);
  }

  /**
   * Create an InputSourceDataInputStream from a byte array
   *
   * @param arr byte array to create from
   * @return InputSourceDataInputStream from a byte array
   */
  public static InputSourceDataInputStream getInputSourceDataInputStream(byte[] arr) {
    return org.apache.daffodil.io.InputSourceDataInputStream.apply(arr);
  }

  /**
   * These are the events that a derived specific InfosetInputter
   * creates.
   * <p>
   * The InfosetInputter base class figures out Daffodil InfosetEvents from
   * the call-backs providing these derived-class events types.
   */
  public enum InfosetInputterEventType {
    StartDocument,
    EndDocument,
    StartElement,
    EndElement;
  }
}
